### DISCUSSION TITLE
How To Center Elements Within Panels

### DISCUSSION THREAD

**Discussion by Lertos** [emilk/egui#1197] [Q&A] [answered]
I have a SidePanel and a CenterPanel. They are working good. Now, inside the CenterPanel I want to split it into two sections - each section with have the items center - both horizontally AND vertically. I would like not to use padding/margin as it doesn't seem you can use percentages - so scaling would be an issue.

Currently all my attempts (using grids, and/or "xx.with_layout" and all its' settings) have only worked for ONE axis. How does one get both to work? Below is a visual example of what I'm looking for:

**Blue** = the group of data I want to center on both axis
**Green** = showing how I want to center each group

![image](https://user-images.githubusercontent.com/19110760/152219880-8d646b4c-5d55-412b-a878-f4b008234203.png)

**Comment by emilk** [comment, marked as answer]
Automatic layouts is somewhere between difficult and impossible with immediate mode UI:s.

In this case: if you know the size of the images, you can do the math and call [`Ui::put`](https://docs.rs/egui/latest/egui/struct.Ui.html#method.put) to manually place the image and text where they belong.

**Reply by Lertos** [reply]
I couldn't agree more! I'm having such a hard time wrapping my head around how complicated UI is - and egui abstracts most of that away! I couldn't imagine the work behind the plugin :P

In my case: it sounds like it might be something where on the resize event we recalculate and "put" our elements where they would need to be?
